core/amber/src/main/scala/edu/uci/ics/amber/engine/common/storage/DatasetFileDocument.scala

```diff
@@ -1,15 +1,24 @@
 package edu.uci.ics.amber.engine.common.storage
 
-import edu.uci.ics.texera.web.resource.dashboard.user.dataset.DatasetResource
+import edu.uci.ics.texera.web.resource.dashboard.user.dataset.service.GitVersionControlLocalFileStorage
+import edu.uci.ics.texera.web.resource.dashboard.user.dataset.utils.PathUtils
+import org.jooq.types.UInteger
 
-import java.io.{File, InputStream, FileOutputStream}
-import java.net.URI
-import java.nio.file.{Files, Path}
+import java.io.{File, FileOutputStream, InputStream}
+import java.net.{URI, URLDecoder}
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Paths}
 
-class DatasetFileDocument(fileFullPath: Path) extends VirtualDocument[Nothing] {
+class DatasetFileDocument(uri: URI) extends VirtualDocument[Nothing] {
+  // Extract path components and decode them
+  private val pathParts = uri.getPath
```

Comment:
this `pathParts` seems to be quite local (only used to calculate `did`, `versionHash` and `relativePath`). In this case, try to avoid making it a class variable. It can be a local variable or lined. 

For example, you can do
```
val did ,versionHash, fileRelativePath = parse(uri) // a function which returns those information 
```

Reply:
changed.